### DISCUSSION TITLE
Validating DITA input without generating output?

### DISCUSSION THREAD

**Discussion by chrispy-snps** [dita-ot/dita-ot#3916] [Q&A] [answered]
Is there a way to run the preprocessing part of the DITA-OT without generating any output, to see if there are any issues (schema, cross-references, missing files, and so on) lurking in the input files?

Conceptually, it would be an "empty" transformation, something like

```
dita --input map.ditamap --format none
                         ^^^^^^^^^^^^^
```

I tried redirecting output to /dev/null:

```
dita --input map.ditamap --format dita --output /dev/null
```

but I get various filesystem call failures with this approach:

```
Attempt to copy experimentdita/iugux/product_installation/graphics/welcome.JPG to /dev/null/iugux/product_installation/graphics/welcome.JPG using NIO Channels failed due to 'failed to create the parent directory for /dev/null/iugux/product_installation/graphics/welcome.JPG'.  Falling back to streams.
     [copy] Failed to copy experimentdita/iugux/product_installation/graphics/welcome.JPG to /dev/null/iugux/product_installation/graphics/welcome.JPG due to java.nio.file.FileSystemException /dev/null/iugux/product_installation/graphics/welcome.JPG: Not a directory

Error: Fatal error during transformation using /home/user/dita-ot/plugins/org.dita.normalize/xsl/normalize.xsl: Failed to create output file file:/dev/nulliugux/preinstallation_setup/finding_your_site_identification_number.dita; SystemID: file:/home/user/dita-ot/plugins/org.dita.normalize/xsl/normalize.xsl; Line#: 91; Column#: 48
```

I guess I could create a custom plugin with a new transformation type that does nothing, but it would be great if there's some clever way to do this with the DITA-OT as-is.

Thanks!

**Comment by jason-fox** [comment]
Would this [DITA Validator Plug-in](https://jason-fox.github.io/dita-ot-plugins/validate.svrl/) suit your purposes?

```console
dita -f svrl -i document.ditamap
```

Once the command has run, all errors and warnings are echoed to the command line

```text
[WARN]   [/out/temp/dita/topics/comment-fixme.dita]
 Line 15: section[id="bad"] - [comment-fixme]
Found 'FIXME' comments within the <section> element - fix as requested and delete the comment.

FIXME This comment requires action

Found 0 Errors 1 Warnings
```

It doesn't rely on the preprocessor since the errors could be within a `*.ditamap` it just operates over all files within a root directory. The plugin was demonstrated as part of [DITA-OT day](https://www.youtube.com/watch?v=GJXB5R9CZyI) in Berlin.

The original DITA-OT 2.x code from HERE can be found [on their GitHub](https://github.com/heremaps/com.here.validate.svrl) a maintained  DITA-OT 3.x fork can be found on [here](https://github.com/jason-fox/com.here.validate.svrl)

**Comment by jelovirt** [comment]
If you only want the validation and checks from preprocess, a custom transtype that only runs preprocess or preprocess2 should do it.

**Comment by chrispy-snps** [comment, marked as answer]
I ended up writing a Perl script that uses `jing` (the RelaxNG validation engine) to do this. Here's a link to the script:

[ditaot_validate.pl](https://github.com/chrispy-snps/DITA-ditaot-utilities#using-ditaot_validatepl)

Because `jing` performs only parsing and validation, it is very fast even for large numbers of files:

![ditaot_validate](https://user-images.githubusercontent.com/50950969/184040212-3278f3f2-4908-4c4c-af47-586cce011fe6.svg)

Note that `ditaot_validate.pl` performs validation of files and directories of files; it does not follow map hierarchy or validate just the files used by a map. For that, the plugin approach described earlier would be better.

Another limitation is that `jing` crashes when validating DITA grammars that include the `svg-d` domain. I am not sure how to resolve this yet.
